Docs: ruler remote evaluation (#1714)
grafana/mimir: docs/sources/operators-guide/architecture/components/ruler/index.md

The ruler is an optional component that evaluates PromQL expressions defined in recording and alerting rules.
Each tenant has a set of recording and alerting rules and can group those rules into namespaces.

## Operational modes

The ruler supports two different rule evaluation modes:

### Internal

This is the default mode. The ruler internally runs a querier and distributor, and evaluates recording and alerting rules in the ruler process itself.
To evaluate rules, the ruler connects directly to ingesters and store-gateways, and writes any resulting series to the ingesters.

Configuration of the built-in querier and distributor uses their respective configuration parameters:

> - [Querier]({{< relref "../../../configuring/reference-configuration-parameters/index.md#querier" >}})

Reviewer: this path is invalid

Reviewer 2: you sure? it worked for me. 
first it failed to load but then it reloaded itself and fetched the page.

Reviewer 3: We also validate all links in CI. If a link is broken (in the doc), the CI doesn't pass.

> this path is invalid

Where are you seeing it broken? I would like to further investigate it.

> ![Architecture of Grafana Mimir's ruler component](ruler.svg)

Reviewer 3: This diagram applies only to the internal operational mode. I would move under "Internal".

## Operational modes

The ruler supports two different rule evaluation modes:

### Internal

This is the default mode. The ruler internally runs a querier and distributor, and evaluates recording and alerting rules in the ruler process itself.
To evaluate rules, the ruler connects directly to ingesters and store-gateways, and writes any resulting series to the ingesters.

Configuration of the built-in querier and distributor uses their respective configuration parameters:

- [Querier]({{< relref "../../../configuring/reference-configuration-parameters/index.md#querier" >}})
- [Distributor]({{< relref "../../../configuring/reference-configuration-parameters/index.md#distributor" >}})

> It should be noted however that when this mode is enabled no acceleration query techniques are performed, meaning that if the rule evaluation complexity is high enough the latency can be increased up to the point of limiting the correct functioning of the component.

Reviewer 3: I would be a bit more specific:

```suggestion
> **Note**: when this mode is used, no query acceleration techniques are used and the evaluation of very high cardinality queries could take longer than the evaluation interval, eventually leading to missing data points in the evaluated recording rules.
```

![Architecture of Grafana Mimir's ruler component](ruler.svg)

## Operational modes

> The ruler supports two different rule evaluation modes, adaptable according to the specific needs of the user:

Reviewer 3: [nit] I would keep it simple:

```suggestion
The ruler supports two different rule evaluation modes:
```

> - **Internal:** This is the default mode. When set, the ruler connects directly through its built-in querier to ingesters and store-gateways, and internally evaluates the rule expression to generate the resulting series.

Reviewer 3: ```suggestion
- **Internal:** This is the default mode. The ruler internally run a querier and distributor, and evaluates recording and alerting rules in the ruler process itself. To evaluate rules, the ruler connects directly to ingesters and store-gateways, and write resulting series to ingesters.
```

Reviewer 4: Edited for grammar

> - **Internal:** This is the default mode. When set, the ruler connects directly through its built-in querier to ingesters and store-gateways, and internally evaluates the rule expression to generate the resulting series.

Reviewer 3: Instead of a list, I would convert it to sub-paragrapths (e.g. `### Internal`). This way we can move the architecture diagram here.

Reviewer 4: I would add a noun after this adjective. Internal [what?]

Reviewer 4: Given that there will be few if any links to the specific modes, let's keep the section as is and we can still add the image(s).

Reviewer 3: The problem is that the diagram above refers only to the "Internal" mode and if we keep it above looks like how the ruler works in any operational mode, which is wrong.

>   In most use cases this mode is solvent enough. However, in those where the evaluation complexity is high enough, the latency can be increased up to the point of limiting the correct functioning of the component.

Reviewer 3: Please clarify query sharding is not used for rule evaluations.

> - **Remote:** In this mode the ruler delegates rule expression evaluation to the query-frontend. In this way, we can leverage all the optimizations that this component offers us, such as [query sharding]({{< relref "../../query-sharding/index.md" >}}), allowing us to horizontally scale and drastically reducing evaluation times.<br/>

Reviewer 3: ```suggestion
- **Remote:** In this mode the ruler delegates rules evaluation to the query-frontend. When enabled, the ruler leverages all the query acceleration techniques employed by the query-frontend, such as [query sharding]({{< relref "../../query-sharding/index.md" >}}).
```

> - **Remote:** In this mode the ruler delegates rule expression evaluation to the query-frontend. In this way, we can leverage all the optimizations that this component offers us, such as [query sharding]({{< relref "../../query-sharding/index.md" >}}), allowing us to horizontally scale and drastically reducing evaluation times.<br/>

Reviewer 3: We need the architecture diagram for this operational mode. @09jvilla who can we ask it?

Reviewer 2: Marco, looks like you've already got your answer :) Song helped me put them together. Let me know if you need more help.

Reviewer 3: I'm good, thanks! Work will be tracked by this issue: https://github.com/grafana/mimir/issues/1831

>   To enable query-frontend rule evaluation, set the `-ruler.query-frontend.address` CLI flag or its respective YAML configuration parameter for the ruler.

Reviewer 3: ```suggestion
  To enable the remote operational mode, set the `-ruler.query-frontend.address` CLI flag or its respective YAML configuration parameter for the ruler.
```

>   Communication between ruler and query-frontend is established over gRPC, so if needed, we can make use of client-side load balancing by prefixing the address value with `dns://`.

Reviewer 3: ```suggestion
  Communication between ruler and query-frontend is established over gRPC, so you can make use of client-side load balancing by prefixing the query-frontend address URL with `dns://`.
```

> ## Recording rules

Reviewer 3: Below, under "Recording rules" it says:

```
The ruler has a built-in querier that evaluates the PromQL expressions and a built-in distributor, so that it can write directly to the ingesters. Configuration of the built-in querier and distributor uses their respective configuration parameters:

- [Querier](http://localhost:3002/docs/mimir/next/operators-guide/configuring/reference-configuration-parameters/#querier)
- [Distributor](http://localhost:3002/docs/mimir/next/operators-guide/configuring/reference-configuration-parameters/#distributor)
```

It should be moved to the new "Internal" section, but it's specific to the "Internal" operational mode.

Reviewer 4: These links are local.

The ruler looks for tenant rules in the `/tmp/rules/<TENANT ID>` directory.
The ruler requires rule files to be in the [Prometheus format](https://prometheus.io/docs/prometheus/latest/configuration/recording_rules/#recording-rules).

## Evaluate rules using query-frontend

> Despite including its own built-in querier, the ruler can be configured to delegate evaluation of rule expressions to the query-frontend.

Reviewer 3: Your doc edit is great. My only take is that if you start reading this doc from the beginning, it's not clear the ruler supports 2 operational modes. I think we should add a paragraph before "Recording rules", explaining the 2 operational modes. I would move the `ruler.svg` under that paragraph and I would work with UI team to create a chart for the new operational mode too.

Author: @pracucci except for the changes in the `ruler.svg` file, I've [included](https://github.com/grafana/mimir/pull/1714/commits/c08cc5fb02e0e58bd38a840048bce60a7b52617e) the new section you suggested, explaining the different operational modes for the ruler. Would be great if you could do the tech review to confirm that everything makes sense.
